[cssom] Serialize the 'all' shorthand in cssText

Loirooriol opened the pull request:
#### 4481e38af645d3bac270241c50053345a9962d8a
<pre>
[cssom] Serialize the &apos;all&apos; shorthand in cssText
<a href="https://bugs.webkit.org/show_bug.cgi?id=190753">https://bugs.webkit.org/show_bug.cgi?id=190753</a>

Reviewed by Darin Adler.

Test: imported/w3c/web-platform-tests/css/cssom/cssstyledeclaration-csstext-all-shorthand.html

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssstyledeclaration-csstext-all-shorthand-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssstyledeclaration-csstext-all-shorthand.html: Added.
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::asTextInternal const):

Canonical link: <a href="https://commits.webkit.org/255329@main">https://commits.webkit.org/255329@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ed06b6c5bef9541fda88e7822bf45f1f0f54c5ad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/92156 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/1386 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/22684 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/101943 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/162121 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/96156 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/1384 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/29777 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/84598 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/98110 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/97813 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/890 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/78679 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/27830 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/82809 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/82408 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/70868 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/36204 "Built successfully") | [⏳ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/macOS-BigSur-Release-WK1-Tests-EWS "Waiting to run tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/33961 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/17538 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3695 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/37831 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/40225 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/39731 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/36665 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1621,6 +1621,19 @@ StringBuilder StyleProperties::asTextInternal() const
         } else {
+            // FIXME: could probably use matchingShorthandsForLonghand() instead of populating 'shorthands' manually.
+            switch (propertyID) {
+            case CSSPropertyCustom:
+            case CSSPropertyDirection:
+            case CSSPropertyUnicodeBidi:
+                // These are the only longhands not included in the 'all' shorthand.
+                break;
```
Given that we had to write out this list by hand, can we write a test that will fail if this list is incorrect? If I remove something from or add something to this list, the test should fail.

Loirooriol replied:
This is covered by the test that I added. I guess the only case that is not tested is if the CSSWG creates another longhand which is not not included in `all`, then it should be added to this list but it won't be detected.

darinadler replied:
How would the test you added fail if I (incorrectly) added `CSSPropertyBorderRadius` to this switch statement?

Loirooriol replied:
Well, that's a shorthand, and we shouldn't be getting shorthands here (even if we do in broken cases). But if for example I add the longhand `CSSPropertyBorderTopLeftRadius` to the switch, then:
```
FAIL 'all' shorthand alone assert_equals: expected "all: inherit;" but got "all: inherit; border-radius: inherit;"
FAIL 'all' shorthand with 'width' and 'height' assert_equals: expected "all: inherit;" but got "all: inherit; border-radius: inherit;"
FAIL 'all' shorthand with 'direction' and 'unicode-bidi' assert_equals: expected "direction: ltr; all: inherit; unicode-bidi: plaintext;" but got "direction: ltr; all: inherit; border-radius: inherit; unicode-bidi: plaintext;"
FAIL 'all' shorthand with 'width', 'height' and custom properties assert_equals: expected "--a: a; all: inherit; --b: b;" but got "--a: a; all: inherit; border-radius: inherit; --b: b;"
FAIL 'all' shorthand with all longhands assert_equals: expected "all: inherit; direction: inherit; unicode-bidi: inherit;" but got "all: inherit; border-radius: inherit; direction: inherit; unicode-bidi: inherit;"
```

Loirooriol replied:
I added one test checking that the only standard longhands not included in `all` are `direction` and `unicode-bidi`. So the CSSWG adds another longhand like that, the test will need to be updated, and then it will fail if Webkit implements the new longhand but forgets to add it into this switch. The test also catches a bug in Blink with `-webkit-border-image`, so it's good.

darinadler replied:
That new test sounds good. Is it already in this pull request, or coming in a future commit?

Loirooriol replied:
Yes, it's already in the pull request:
```js
test(function() {
  let cssText = "all: inherit; ";
  for (let longhand of getComputedStyle(document.documentElement)) {
    cssText += longhand + ": inherit; ";
  }
  style.cssText = cssText;
  assert_equals(style.cssText, "all: inherit; direction: inherit; unicode-bidi: inherit;");
}, "'all' shorthand with all longhands");
```

darinadler commented on Source/WebCore/css/StyleProperties.cpp:
```diff
@@ -1621,6 +1621,19 @@ StringBuilder StyleProperties::asTextInternal() const
+                break;
+            default:
+                ASSERT(propertyID >= firstCSSProperty);
+                ASSERT(propertyID < firstShorthandProperty);
```
Looks like this assertion is failing in many tests; need to figure that out before landing this.

Loirooriol replied:
Yeah, something like this:

```html
<!DOCTYPE html>
<textarea></textarea>
<script>
var input = document.querySelector('textarea');
input.focus();
input.selectionStart = 0;
document.execCommand('InsertHTML', false, '.');
</script>
```

somehow sets `animation: 0s ease 0s 1 normal none running none; background: rgba(0, 0, 0, 0) none repeat scroll 0% 0% / auto padding-box border-box; background-position: 0% 0%; border: 0px none rgb(0, 0, 0); border-block: 0px none rgb(0, 0, 0); border-block-color: rgb(0, 0, 0)�; border-block-end: 0px none rgb(0, 0, 0); border-block-start: 0px none rgb(0, 0, 0); border-block-style: none; border-block-width: 0px; border-bottom: 0px none rgb(0, 0, 0); border-color: rgb(0, 0, 0); border-image: none; border-inline: 0px none rgb(0, 0, 0)
                                                                                                                                                                                                                                                            �6; border-inline-color: rgb(0, 0, 0)�; border-inline-end: 0px none rgb(0, 0, 0); border-inline-start: 0px none rgb(0, 0, 0)
                                                                                                         �6; border-inline-style: noneal; border-inline-width: 0px; border-left: 0px none rgb(0, 0, 0)
                                                                                                                                                                                                      �6; border-radius: 0pxlace; border-right: 0px none rgb(0, 0, 0)
                                                                                                                                                                                                                                                                     �6; border-spacing: 0px 0px; border-style: none; border-top: 0px none rgb(0, 0, 0); border-width: 0px; column-rule: 0px none rgb(0, 0, 0); columns: auto auto; container: none; flex: 0 1 auto; flex-flow: row nowrap; font: normal normal normal normal 13px/normal "Noto Sans"; font-variant: normal; gap: normal normal; grid: none / none / none / row / auto / auto; grid-area: auto / auto / auto / auto���; grid-column: auto / auto; grid-row: auto / auto; grid-template: none / none / none; inset: auto; inset-block: auto; inset-inline: auto; list-style: outside none disc; margin: 0px; margin-block: 0px; margin-inline: 0px; mask: none; mask-position: 0% 0%; offset: auto none 0px auto 0deg / auto; outline: rgb(0, 0, 0) none 0px; overflow: visible; padding: 0px; padding-block: 0px; padding-inline: 0px; page-break-after: auto; page-break-before: auto; page-break-inside: auto; perspective-origin: 80.5px 9px; place-content: normal normal; place-items: normal normal; place-self: auto auto; scroll-margin: 0px; scroll-margin-block: 0pxo 0deg; scroll-margin-inline: 0px; scroll-padding: auto; scroll-padding-block: auto; scroll-padding-inline: auto; text-decoration: none; text-decoration-skip: auto; text-emphasis: none rgb(0, 0, 0); transform-origin: 80.5px 9px; transition: all 0s ease 0s; -webkit-background-size: auto; -webkit-border-image: none; -webkit-column-break-after: auto; -webkit-column-break-before: auto; -webkit-column-break-inside: auto; -webkit-mask-position: 0% 0%; -webkit-perspective: none; -webkit-text-decoration: none solid rgb(0, 0, 0)�; -webkit-text-orientation: mixed;` without expanding them into longhands!
This seems pretty broken but I guess it's unrelated. I guess I can just add a conditional with a FIXME.

Loirooriol replied:
The reason for the assert failure was that `ComputedStyleExtractor::copyProperties()` copies both the longhands and the shorthands (without expanding them). This seems a regression from 7ee771049e3b77c5fa5f40ad9acc7703d6557c03, since previously in only copied computable properties, which are longhands.

darinadler replied:
Good call; let’s fix that regression, too, maybe in a separate patch, and then we can add back the assertion. Attn: @dcrousso and @smfr.

Loirooriol replied:
Created https://github.com/WebKit/WebKit/pull/5157

Loirooriol committed:
[cssom] Serialize the 'all' shorthand in cssText
https://bugs.webkit.org/show_bug.cgi?id=190753

Reviewed by Darin Adler.

Test: imported/w3c/web-platform-tests/css/cssom/cssstyledeclaration-csstext-all-shorthand.html

* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssstyledeclaration-csstext-all-shorthand-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssstyledeclaration-csstext-all-shorthand.html: Added.
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::asTextInternal const):

Canonical link: https://commits.webkit.org/255329@main
